# System features (CPU and GPU)

In [11]:
!sudo apt-get install neofetch
!neofetch

Reading package lists... Done
Building dependency tree       
Reading state information... Done
neofetch is already the newest version (3.4.0-1).
0 upgraded, 0 newly installed, 0 to remove and 11 not upgraded.
            .-/+oossssoo+/-. 
        `:+ssssssssssssssssss+:` 
      -+ssssssssssssssssssyyssss+- 
    .ossssssssssssssssssdMMMNysssso. 
   /ssssssssssshdmmNNmmyNMMMMhssssss/ 
  +ssssssssshmydMMMMMMMNddddyssssssss+ 
 /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/ 
.ssssssssdMMMNhsssssssssshNMMMdssssssss. 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+ 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+ 
.ssssssssdMMMNhsssssssssshNMMMdssssssss. 
 /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/ 
  +sssssssssdmydMMMMMMMMddddyssssssss+ 
   /ssssssssssshdmNNNNmyNMMMMhssssss/ 
    .ossssssssssssssssssdMMMNysssso. 
      -+sssssssssssssssssyyyssss+- 
        `:+ssssssssssssssssss+:` 
            .-/+oossssoo+/-. 
root@a1638f08f800 
---

In [12]:
!nvidia-smi

Mon Nov  9 13:49:04 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 455.32.00    Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   51C    P8    10W /  70W |      0MiB / 15079MiB |      0%      Default |
|                               |                      |                 ERR! |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [13]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0x1
cpu MHz		: 2200.000
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa
bogomips	: 4400.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 b

## Test 1

In [15]:
import time, torch

bas = time.time()

a = torch.rand(10000, 10000, device=torch.device("cuda"))
b = torch.rand(10000, 10000, device=torch.device("cuda"))
c = torch.rand(10000, 10000, device=torch.device("cuda"))
d = torch.rand(10000, 10000, device=torch.device("cuda"))


for i in range(0,1000):
  y = a@b
  z = c@d

  x = y@z

son = time.time()

In [65]:
x

tensor([[6.3077e+10, 6.3017e+10, 6.2942e+10,  ..., 6.3034e+10, 6.2659e+10,
         6.3672e+10],
        [6.1768e+10, 6.1709e+10, 6.1636e+10,  ..., 6.1726e+10, 6.1359e+10,
         6.2350e+10],
        [6.1629e+10, 6.1570e+10, 6.1498e+10,  ..., 6.1587e+10, 6.1221e+10,
         6.2211e+10],
        ...,
        [6.2017e+10, 6.1957e+10, 6.1884e+10,  ..., 6.1974e+10, 6.1606e+10,
         6.2601e+10],
        [6.3193e+10, 6.3133e+10, 6.3058e+10,  ..., 6.3150e+10, 6.2775e+10,
         6.3789e+10],
        [6.2905e+10, 6.2844e+10, 6.2770e+10,  ..., 6.2861e+10, 6.2488e+10,
         6.3498e+10]], device='cuda:0')

In [20]:
"1.test result (second) : " + str(son-bas)

'1.test result (second) : 1293.739860534668'

## Test 2

In [21]:
!nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243


In [22]:
!pip install git+git://github.com/andreinechaev/nvcc4jupyter.git
%load_ext nvcc_plugin

  Cloning git://github.com/andreinechaev/nvcc4jupyter.git to /tmp/pip-req-build-1mcug2nf
  Running command git clone -q git://github.com/andreinechaev/nvcc4jupyter.git /tmp/pip-req-build-1mcug2nf
  Created wheel for NVCCPlugin: filename=NVCCPlugin-0.0.2-cp36-none-any.whl size=4307 sha256=9bb504d4afa9e1e8145b439dcd79fa4c048c7f7fca42eda5bdd08e22cf9c4534
  Stored in directory: /tmp/pip-ephem-wheel-cache-7ett1rf2/wheels/10/c2/05/ca241da37bff77d60d31a9174f988109c61ba989e4d4650516
Successfully built NVCCPlugin
created output directory at /content/src
Out bin /content/result.out


In [34]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/kernel.cu

--2020-11-09 14:36:11--  https://ahmetfurkandemir.s3.amazonaws.com/kernel.cu
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.216.111.11
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.216.111.11|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1155 (1.1K) [application/cu-seeme]
Saving to: ‘kernel.cu’

kernel.cu           100%[===================>]   1.13K  --.-KB/s    in 0s      

2020-11-09 14:36:11 (74.4 MB/s) - ‘kernel.cu’ saved [1155/1155]



In [35]:
!cat dev_array.h

#ifndef _DEV_ARRAY_H_
#define _DEV_ARRAY_H_

#include <stdexcept>
#include <algorithm>
#include <cuda_runtime.h>

template <class T>
class dev_array
{
// public functions
public:
    explicit dev_array()
        : start_(0),
          end_(0)
    {}

    // constructor
    explicit dev_array(size_t size)
    {
        allocate(size);
    }
    // destructor
    ~dev_array()
    {
        free();
    }

    // resize the vector
    void resize(size_t size)
    {
        free();
        allocate(size);
    }

    // get the size of the array
    size_t getSize() const
    {
        return end_ - start_;
    }

    // get data
    const T* getData() const
    {
        return start_;
    }

    T* getData()
    {
        return start_;
    }

    // set
    void set(const T* src, size_t size)
    {
        size_t min = std::min(size, getSize());
        cudaError_t result = cudaMemcpy(start_, src, min * sizeof(T), cudaMemcpyHostToDevice);
        if (result != cudaSuccess)
        {
     

In [38]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/dev_array.h

--2020-11-09 14:37:28--  https://ahmetfurkandemir.s3.amazonaws.com/dev_array.h
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.217.104.228
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.217.104.228|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1997 (2.0K) [text/x-chdr]
Saving to: ‘dev_array.h’

dev_array.h         100%[===================>]   1.95K  --.-KB/s    in 0s      

2020-11-09 14:37:29 (124 MB/s) - ‘dev_array.h’ saved [1997/1997]



In [39]:
!cat dev_array.h

#ifndef _DEV_ARRAY_H_
#define _DEV_ARRAY_H_

#include <stdexcept>
#include <algorithm>
#include <cuda_runtime.h>

template <class T>
class dev_array
{
// public functions
public:
    explicit dev_array()
        : start_(0),
          end_(0)
    {}

    // constructor
    explicit dev_array(size_t size)
    {
        allocate(size);
    }
    // destructor
    ~dev_array()
    {
        free();
    }

    // resize the vector
    void resize(size_t size)
    {
        free();
        allocate(size);
    }

    // get the size of the array
    size_t getSize() const
    {
        return end_ - start_;
    }

    // get data
    const T* getData() const
    {
        return start_;
    }

    T* getData()
    {
        return start_;
    }

    // set
    void set(const T* src, size_t size)
    {
        size_t min = std::min(size, getSize());
        cudaError_t result = cudaMemcpy(start_, src, min * sizeof(T), cudaMemcpyHostToDevice);
        if (result != cudaSuccess)
        {
     

In [40]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/kernel.h

--2020-11-09 14:38:22--  https://ahmetfurkandemir.s3.amazonaws.com/kernel.h
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.216.101.123
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.216.101.123|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 112 [text/x-chdr]
Saving to: ‘kernel.h’

kernel.h            100%[===================>]     112  --.-KB/s    in 0s      

2020-11-09 14:38:22 (3.23 MB/s) - ‘kernel.h’ saved [112/112]



In [41]:
!cat kernel.h

#ifndef KERNEL_CUH_
#define KERNEL_CUH_

void matrixMultiplication(float *A, float *B, float *C, int N);

#endif

In [82]:
!wget https://ahmetfurkandemir.s3.amazonaws.com/matrixmul.cu

--2020-11-09 15:07:23--  https://ahmetfurkandemir.s3.amazonaws.com/matrixmul.cu
Resolving ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)... 52.216.95.35
Connecting to ahmetfurkandemir.s3.amazonaws.com (ahmetfurkandemir.s3.amazonaws.com)|52.216.95.35|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1060 (1.0K) [application/cu-seeme]
Saving to: ‘matrixmul.cu’

matrixmul.cu        100%[===================>]   1.04K  --.-KB/s    in 0s      

2020-11-09 15:07:23 (44.7 MB/s) - ‘matrixmul.cu’ saved [1060/1060]



In [83]:
!cat matrixmul.cu

#include <iostream>
#include <vector>
#include <stdlib.h>
#include <time.h>
#include <cuda_runtime.h>
#include "kernel.h"
#include "kernel.cu"
#include "dev_array.h"
#include <math.h>
#include <stdio.h>

using namespace std;

int main()
{
    // Perform matrix multiplication C = A*B
    // where A, B and C are NxN matrices
    int N = 10000;
    int SIZE = N*N;

    // Allocate memory on the host
    vector<float> h_A(SIZE);
    vector<float> h_B(SIZE);
    vector<float> h_C(SIZE);

    // Initialize matrices on the host
    for (int i=0; i<N; i++){
        for (int j=0; j<N; j++){
            h_A[i*N+j] = sin(i);
            h_B[i*N+j] = cos(j);
        }
    }

    // Allocate memory on the device
    dev_array<float> d_A(SIZE);
    dev_array<float> d_B(SIZE);
    dev_array<float> d_C(SIZE);

    d_A.set(&h_A[0], SIZE);
    d_B.set(&h_B[0], SIZE);

    matrixMultiplication(d_A.getData(), d_B.getData(), d_C.getData(), N);
    cudaDeviceSynchronize();

    d_C.get(&h_C[0], SIZE);
    c

In [86]:
# compilation test
bas = time.time()
!nvcc matrixmul.cu
son = time.time()

In [87]:
"2. test - compilation test result : " + str(son-bas)

'2nd test - compilation test result : 1.6754465103149414'

In [88]:
# run the compiled file, test
bas = time.time()
!./a.out
son = time.time()

END

In [90]:
"2. test - time to run the compiled file : "+str(son-bas)

'2. test - time to run the compiled file : 8.99604868888855'

**1. Test Result : 1293.739860534668 (second)**

**2. Test Reult - Compile : 1.6754465103149414 (second)**

**2. Test Reult - run file : 8.99604868888855 (second)**